### ISSUE TITLE
SynchronizedObservableCollection RaceCondition

### ISSUE THREAD

**Issue by Andrew-Hanlon** [CoryCharlton/CCSWE-Libraries#2]
I have been using a modified version of your SynchronizedObservableCollection for some time, but have recently came across an issue with the the combination of Property and Collection change notifications being "dispatched" (pushed onto the context) separately. 

In many of the methods the notifications are ordered as:

OnPropertyChanged("Count");
OnPropertyChanged("Item[]");
OnCollectionChanged(NotifyCollectionChangedAction.Remove, value, index);

Which is exactly the order from the ObservableCollection source; however since here the property and collection changes are being dispatched separately, there is a possibility of other operations being pushed onto the context in-between those notifications (a semi race condition). In some cases this means that a bound control (I am using WPF) responds to the property change notification first and then processes the collection change separately - effectively updating its state twice. This can lead to very strange behaviour such as two duplicate items being added to a UI itemspanel after a single Add. 

The fix is simply to combine the property and collection notifications into a single dispatch operation - which is a good performance idea anyway. My attached code has a number of modifications that are likely not quite as universally applicable, but you can see how the notifications have been grouped into specific methods and are dispatched together.

Thank you and kind regards!

[SynchronizedObservableCollection.zip](https://github.com/CoryCharlton/CCSWE-Libraries/files/253891/SynchronizedObservableCollection.zip)

**Comment by CoryCharlton** [comment]
Sorry for the delayed response, somehow this slipped passed me. Really appreciate your feedback and providing code on how you resolved this issue. I'll see if I can duplicate your implementation while retaining the `SynchronizationContext` for compatibility.

**Comment by Andrew-Hanlon** [comment]
Not to worry. In fact, since writing that message I have completely changed my own approach. I found that by using WPF's BindingOperations.EnableCollectionSynchronization() method and BindingOperations.CollectionRegistering event in .net 4.5, I was able to write a simpler Collection that requires no dispatching/context switching, and self synchronizes only when needed. The EnableCollectionSync method even has an overload to allow you to provide your own synchronization wrapper, so the ReaderWriter can still be used (albeit I've moved away from it due to the unmanaged IDisposable requirement).

I still use the above method in certain scenarios where I'm using third party controls that have custom/broken CollectionView implementations that don't seem to play nicely with 4.5.  I'll share some code soon. Cheers!
